### ISSUE TITLE
metaserver development

### ISSUE THREAD

**Issue by mimooh** [Wargus/stratagus#194]
Metaserver allows for playing internet games: players announce new games and
then other clients can connect to those games. There's some C code in stratagus
which is not finished - I cannot judge how complete this code is. Unless
there's a chance/person willing to continue this work, I could try with a new
architecture from scratch (I don't know C). My idea is as follows:

A php/sql site where users can announce new games. It can have a lobby added,
where we choose map/race/resources and such (thanks to timfel we can pass these
parameters to wargus via command line, so we don't need to go thru the lobby in
wargus anymore). Once we are done with that, the game participants enter a P2P
network and they talk to each other directly since then. The P2P would be a
python framework, such as https://pypi.python.org/pypi/pyp2p, which I hope I
could handle. Python and dependency modules would need to be installed on all
clients. 

"Pyp2p is designed to work with simple non-blocking TCP sockets. To use them,
your application must create an infinite loop which is used to periodically
look for new replies. As you look for replies, the software also handles
accepting new connections and removing old ones automatically.

Pyp2p is a simplified networking library for building peer-to-peer networks in
Python. The library is designed to solve the pain of finding nodes and
bypassing NATs so you can focus on writing your application code."

P2P clients would exchange the messages via network on one end and with
stratagus on the other end. So, on this second end, either P2P is hardwired to
stratagus or they would be separate processes exchanging data via pipes or
similar. Both requires some stratagus code changes for integration, by someone... 
And this someone is crucial, otherwise my work would be useless.

What do you think of this idea? You know, the code is made in C and then there
are those few people that are good enough to make further development and we
are stuck. Seems like more people are capable of Lua/Python/PHP to make the
fixes so I prefer such languages where they are sufficient. 

@Mistranger, are you around? You developed lot's of stratagus and we need your
advice on the metaserver and these spells issues that were brought up lately.

**Comment by Mistranger** [comment]
Sorry, guys, I left this because this engine needs serious refactoring, well, no, it needs to be rewritten from scratch. There is a lot of code which hasn't updated itself since FreeCraft project, it's a big mess of optimized C code and some Jarod42's C++ refactoring.
Still, I would recommend to use https://github.com/OpenRA/OpenRA - a good example of RTS engine, which is actively maintained, has perfect internal structure and it has a good multiplayer/metaserver support with support for dedicated servers already. One day I have started to move on my Doom Wars project on it - https://github.com/Mistranger/OpenRA. And don't stay on Stratagus, unless you really want it and have a lot of time to work on it (I've spent around 5 years on it, it's a surprise how Andrettin haven't gave up yet, heh)

**Comment by DinkyDyeAussie** [comment]
Oooo C# used for OpenRA? Hmmm might have to hop onboard this. If we can get things done for wargus on the OpenRA engine that would be fantastic!

**Comment by mimooh** [comment]
@Mistranger, thanks for all the hints and comments. Never heard of OpenRA before. Do you think Wargus could use OpenRA? I guess it would be mostly lua scripting to integrate the two?

**Comment by Erenussocrates** [comment]
Well I would have happily got onboard with this if OpenRA already had a warcraft 2 project on them. If someone decides/starts to do this, I can happily join. I'm not even sure if anyone from their side would bother since they would think that this already exists for stratagus

**Comment by DinkyDyeAussie** [comment]
Wargus would work great on OpenRA! I would join straight away. The code is very neat and lua scripting would change but not too much. I haven't looked into the code in the scripts but I have had a look ar the engine source.

Very nice code in there and C# is way easier than C++. I think it's time for a full reboot of this engine in a more managable and modern language.

**Comment by Mistranger** [comment]
In both cases there would be a lot of work to be done: OpenRA mostly focuses on old C&C games (C&C itself, Red Alert, Dune 2000 and Tiberian Sun). Someone should add support for Warcraft 2 style production, worker class that builds buildings, second resource type, spell casting support and so on.
This someone should also write support for Warcraft 2 internal formats for OpenRA and make a new installer. But OpenRA allows you to do such things without hardcoding the engine (as Stratagus is made), every game based on it is actually a mod, which runs on engine. Look at how they implement game specific features (in case of Dune 2000 - worms, pavement system and so on), they don't add anything to the engine itself. 
In case of Stratagus - you have everything ready, you just need to add this metaserver support, but look - every new feature produces a lot of bugs because of how Stratagus internal structure is made (because no one have rewrote it since 1999).
And yes - I won't be that "someone", I have already done a lot of things to Stratagus and won't touch that thing again.

**Comment by Andrettin** [comment]
I don't really see the point in moving to OpenRA, to be honest. Stratagus does what it is supposed to pretty well, and it would take a massive effort to move to the OpenRA engine. Furthermore, since their license isn't compatible with Stratagus' (as they use GPLv3 while Stratagus uses GPLv2), it wouldn't be possible to reuse any of the existing code.

I think it would be more sensible to rewrite the multiplayer code in Stratagus; the multiplayer desyncs are pretty much the only significant issue the engine has.

**Comment by mimooh** [comment]
Good to hear that Andrettin! :) I don't experience what I would call
problems with stratagus. I'd say it's pretty much complete and working
good, despite internal mess perhaps. 

Could you comment on what I was suggesting for metaserver? If some
other framework (python?) could handle the outside-LAN-multiplayer
(organizing the P2P, firewalls bypassing and website for
announcing/joining games), do you think it could be connected to
stratagus core? Otherwise the gaming over the internet is pretty much
restricted.

-- 
Karol Kreński
tel. 517-302-055

**Comment by Andrettin** [comment]
Yeah, P2P is quite restrictive =/ It should be possible for another framework to handle those things, yes. Do you have something specific in mind?

**Comment by mimooh** [comment]
As I said before: https://pypi.python.org/pypi/pyp2p. Or any other,
because I have no experience with python networking at all, but I know
enough of networking hopefully.

This sounds crucial for how stratagus and pyp2p should be integrated:

"PyP2P is designed to work with simple non-blocking TCP sockets. To use
them, your application must create an infinite loop which is used to
periodically look for new replies"

-- 
Karol Kreński
tel. 517-302-055

**Comment by Mistranger** [comment]
Okay, I won't cross against you, I just want to mention that by moving to OpenRA you will upgrade the codebase. Stratagus was originally intended to be a Warcraft 2 clone, and since no one have updated the codebase, no other games could be built on it without serious limitations (that's why Stargus isn't being developed).
As for license, GPL3 was created in 2007 and Stratagus in 1999, so I think it's possible to upgrade.
Network code isn't the only flaw Stratagus has: hacky and primitive terrain code, bad Lua integration (the one fault I've done here is extending animations scripting instead of moving to Lua), bad error handling (use debugger every time in case of non-Lua errors), "Warcraft 2"-style grid and movement system and mech, much more, so it's not even possible to recreate Starcraft on it without serious refactoring. So moving to OpenRA will get you a lot of advantages with the only drawback - to implement Warcraft 2 on it. And yes, in both cases there will be a lot of work, and of course it shouldn't be done in "hackish" way.

**Comment by timfel** [comment]
For simple UDP hole punching for quick matches over the internet, I've updated the metaserver code and lua integration. It's on a branch right now, but good enough to just have the metaserver running somewhere, getting game listings and connecting remotely from different clients in different networks. Since we hacked dedicated server support into the engine last year, it should be easy to just have a script on the server that also launches dedicated stratagus servers and advertises them to the world in this way, with map cycling and stuff. But since all that can be done outside of the engine with a few lines of bash, I don't think that should go in there, too.

**Comment by mimooh** [comment]
@timfel, you are simply crazy! :) Can't believe how easily you handle such non-trivial stuff. And how could I have missed that wargus can be already run as dedicated - partly because it's under man wargus and not under wargus --help. I will improve the documentation. 

I'll setup the deciated wargus server then and whoever is best player in github wargus community I chalenge you in playing wargus! :)

Once metaserver is ready for registering/joining client games I can help with the php server and perhaps run one myself.

**Comment by Erenussocrates** [comment]
I would take that challenge up myself, as soon as the Wargus multiplayer is ready to use without any problems

**Comment by mimooh** [comment]
I will try to setup this, perhaps tonight. You will need to use v2.4.1.

-- 
Karol Kreński
tel. 517-302-055

**Comment by mimooh** [comment]
@timfel, I am testing the dedicated server:

1. It starts with the GUI. I would expect it should start with no GUI, to ease launching it on a remote server. 

2. Launching the dedicated server also has me connected as the first player. I would expect the server to have 0 players at this moment and wait for players to join.

The command I used:
wargus -c multiplayer -G server,numplayers=2,dedicated,race=orc,resources=high,fow=0,reveal=1,player=$HOSTNAME,map='./maps/skirmish/(2)mysterious-dragon-isle.smp'

@Erenussocrates, seems like we need to wait a while until this is resolved.

**Comment by timfel** [comment]
@mimooh: seems I need to look at the dedicated stuff again. But if all you want is play against someone yourself (rather than hosting a dedicated server for other players), all you need to launch on the remote server is the metaserver (`metaserver -p` gives you some debug output). Then put the address of your metaserver into your preferences.lua file and, when you create a game, there should be a tick mark next to "Advertise this server online". If it says that the metaserver is down or has an error, check the server log. Anyone who wants to join needs to put in the same metaserver into their preferences.lua (sorry, no gui yet). Then use "Join Online Game" to see a list of games waiting to start and connect to them.

**Comment by mimooh** [comment]
I can and even prefer to launch a standalone dedicated metaserver. So
metaserver is a separate binary? It's not part of my stratagus debian
package.

-- 
Karol Kreński
tel. 517-302-055

**Comment by timfel** [comment]
Yes, it's not packaged right now. I'll do that sometime next week.

mimooh <notifications@github.com> schrieb am Sa., 12. Nov. 2016, 14:31:

**Comment by mimooh** [comment]
I decided to sponsor a VPS server, mainly for running the upcoming metaserver.  It is ubuntu 16.04 with root access. If anyone would like to get the access to that machine you can request accounts at mimoohowy@gmail.com. 

I expect once the metaserver is developed as a backend then we can setup the frontend part with a website for announcing/joining wargus games. If anyone cares about rankings we could have a database with users scores.

**Comment by mimooh** [comment]
Now I am curious about launching the metaserver service on that new
server. Metaserver installs as part of stratagus now which is great
news! :) Still I don't quite understand most of metaserver -h and cannot
install wargus, but once I learn how it works we can finally have the
multiplayer service launched.

So how does it operate? One metaserver instance serves only one game to
say 2 players? Or it serves multiple such 2 player games at the same
time?

In ~/.stratagus/preferences.lua I can see:
wc2.preferences.MetaServer = "localhost"
How do I specify port?

**Comment by mimooh** [comment]
@timfel, the metaserver development and the installation process are going pretty well! My quick observations for now are:
1. I don't have videos in my warcraft2 dataset and after the installation via this new nice dialog all videos were of size 0, which prevented wargus from launching. After rm -rf videos/* wargus started and was fine.
2. Under windows, after the installation and extraction, on the desktop there are Aleona and Wargus icons. But Wargus launches Aleona game. Didn't have time to inspect that. 
3. Then I played with the new Metaserver menu. After a bit of experimenting I managed to launch the remote metaserver, have the first client announce the game, have the second client see the game in the listing, but 'connect' didn't make it into the game. Which is quite far I would say anyway. Good work!
